src/m365/spo/commands/file/file-roleassignment-add.ts

```diff
@@ -0,0 +1,270 @@
+      },
+      {
+        option: '--principalId [principalId]'
+      },
+      {
+        option: '--upn [upn]'
+      },
+      {
+        option: '--groupName [groupName]'
+      },
+      {
+        option: '--roleDefinitionId [roleDefinitionId]'
+      },
+      {
+        option: '--roleDefinitionName [roleDefinitionName]'
+      }
+    );
+  }
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
+        const isValidSharePointUrl: boolean | string = validation.isValidSharePointUrl(args.options.webUrl);
+        if (isValidSharePointUrl !== true) {
+          return isValidSharePointUrl;
+        }
+
+        if (args.options.fileId && !validation.isValidGuid(args.options.fileId)) {
+          return `${args.options.fileId} is not a valid GUID`;
+        }
+
+        if (args.options.principalId && isNaN(args.options.principalId)) {
+          return `Specified principalId ${args.options.principalId} is not a number`;
+        }
+
+        if (args.options.roleDefinitionId && isNaN(args.options.roleDefinitionId)) {
+          return `Specified roleDefinitionId ${args.options.roleDefinitionId} is not a number`;
+        }
+
+        const principalOptions: any[] = [args.options.principalId, args.options.upn, args.options.groupName];
+        if (!principalOptions.some(item => item !== undefined)) {
+          return `Specify either principalId, upn or groupName`;
+        }
+
+        if (principalOptions.filter(item => item !== undefined).length > 1) {
+          return `Specify either principalId, upn or groupName but not multiple`;
+        }
+
+        const roleDefinitionOptions: any[] = [args.options.roleDefinitionId, args.options.roleDefinitionName];
+        if (!roleDefinitionOptions.some(item => item !== undefined)) {
+          return `Specify either roleDefinitionId or roleDefinitionName`;
+        }
+
+        if (roleDefinitionOptions.filter(item => item !== undefined).length > 1) {
+          return `Specify either roleDefinitionId or roleDefinitionName but not multiple`;
+        }
+
+        return true;
+      }
+    );
+  }
+
+  #initOptionSets(): void {
+    this.optionSets.push(['fileId', 'fileUrl']);
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    if (this.verbose) {
+      logger.logToStderr(`Adding role assignment to file in site at ${args.options.webUrl}...`);
+    }
+
+    try {
+      const fileUrl: string = await this.getFileURL(args);
+      args.options.roleDefinitionId = await this.getRoleDefinitionId(args.options);
+      if (args.options.upn) {
+        args.options.principalId = await this.getUserPrincipalId(args.options);
+        await this.addRoleAssignment(fileUrl, args.options, logger);
+      }
+      else if (args.options.groupName) {
+        args.options.principalId = await this.getGroupPrincipalId(args.options);
+        await this.addRoleAssignment(fileUrl, args.options, logger);
```

Comment:
it's a mistake also I did so sorry if you copied/pasted it (I think we should open an issue to correct it in existing commands 🤔). 
Generally, we should not overwrite `args` and consider them as read only. let's create a local variable to keep the `principalId`